src/librustc_codegen_ssa/back/linker.rs

```diff
@@ -389,10 +406,11 @@ impl<'a> Linker for GccLinker<'a> {
             return
         }
 
-        // Symbol visibility takes care of this for the WebAssembly.
+        // Symbol visibility takes care of this for the WebAssembly & uwp.
         // Additionally the only known linker, LLD, doesn't support the script
         // arguments just yet
-        if self.sess.target.target.arch == "wasm32" {
+        if self.sess.target.target.arch == "wasm32" ||
+           self.sess.target.target.target_vendor == "uwp" {
```

Comment:
How come this is needed? WebAssembly is a special case here but shouldn't UWP behave like MinGW here? What happened if this clause wasn't here?

Reply:
That's because llvm linker doesn't support `--version-script` when targetting PE/COFF

My understanding is that the format itself doesn't really support versioning, causing LLVM people to not support the flag.
I agree that this is not pretty, but I'm not sure of any other way to fix that

Reply:
Hm how come this isn't getting caught by an existing check for `pc-windows-gnu`? Or rather this should have the same logic for being skipped as `pc-windows-gnu` does.

Reply:
GNU linker accepts the flag, but I'm not sure it does anything useful with it. In any case the flag is passed when building for `pc-windows-gnu`

Reply:
Is the mingw linker for uwp not the gnu linker? Is it a separate path in the code that rejects `--version-script`? Is there perhaps an alternate option that works on both that we could use?

Reply:
If it's an issue between the gnu linker and the llvm linker, then there should be a check based on which linker is being used, not which target is being targeted.

Reply:
Even recent mingw-w64 GCC doesn't work very well for UWP, leave alone the old version used by Rust.
The toolchain which this target would use lives in [this repository](https://github.com/mstorsjo/llvm-mingw), it's mingw-64 + full LLVM (clang, compiler-rt, lld, libc++, ...).

Reply:
I agree that the correct fix should be detecting the linker type. Currently I also need this patch to build the pc-windows-gnu toolchain, since I use the llvm-mingw toolchain for both.
I'll try to work something out and will update the PR

Reply:
I'm currently trying to fix this, but while it was easy to detect that `<arch>-w64-mingw32-gcc` is actually clang (for now this is done in the `cc-rs` crate), I'm having troubles forwarding that information to the rust build system.

AFAICS, only the bootstrap crate uses the `cc::Tool` structure, which contains the valuable bits about the compiler.
I thought about forward this to the `spec::Target` struct, but it seems this is mostly meant for information that are contained in the various specs files, not for a "runtime" information about the compiler that's being used.
The `config::Options` seems to be meant for CLI options, and would therefor not be suited for an information about the compiler.
The last option I'm seing is to store that directly into the `Session` itself, but I'm reluctant in sprinkling booleans here and there.

I'd appreciate any pointer to the correct way of letting various crates (or in this case, the `librustc_codegen_ssa` one) about the actual underlying compiler, as I'm kind of hitting a wall for now

Reply:
Although we can do so I don't think we'll want to do detection of gcc/clang here. I was unaware that the UWP toolchain was clang-based (or that it was even possible to have a clang-based mingw toolchain). If that's the case then this is being added because the linker is LLD and LLD continues to not support this. I think that's fine, but could comments be added in the target specifications that although the linker is specified as gcc it's actually LLD under the hood?